File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        int[] prefix = new int[n];
        int[] postfix = new int[n];
        
        int p = 1;
        for(int i = 0; i<n; i++){
            p *= nums[i];
            prefix[i] = p;
        }

        int q = 1;
        for(int i = n-1; i>0; i--){
            q *= nums[i];
            postfix[i] = q;
        }
        
        for(int i = 0; i<n; i++){
            if(i==0) res[i] = postfix[i+1];
            else if (i==n-1) res[i] = prefix[i-1];
            else{
                res[i]= prefix[i-1]*postfix[i+1];
            }
        }

        return res;
        
    }
}  

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        n = len(nums)
        prefixes = [1] * n
        suffixes = [1] * n
        answer = []

        for i in range(1,n,1):
            prefixes[i] = nums[i-1]*prefixes[i-1]
        for i in range(n-2,-1,-1):
            suffixes[i] = nums[i+1]*suffixes[i+1]
        for i in range(n):
            answer.append(prefixes[i]*suffixes[i])
        
        return answer